src/main/drivers/compass/compass_hmc5883l.c

```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+ * Nevertheless, it is nice to conform to the datasheet specification when per device SPI mode
+ * is implemented.
+ */
+static void hmc5883SpiInit(busDevice_t *busdev)
+{   
+    static bool hardwareInitialised = false;
+
+    if (hardwareInitialised) {
+        return;
+    }
+
+    IOInit(busdev->busdev_u.spi.csnPin, OWNER_COMPASS_CS, 0);
+    IOConfigGPIO(busdev->busdev_u.spi.csnPin, IOCFG_OUT_PP);
+
+    IOHi(busdev->busdev_u.spi.csnPin); // Disable
+
+    spiSetDivisor(busdev->busdev_u.spi.instance, SPI_CLOCK_STANDARD);
+
+    hardwareInitialised = true;
+}
 #endif
+
+static bool hmc5883lRead(magDev_t *magdev, int16_t *magData)
+{
+    uint8_t buf[6];
+
+    busDevice_t *busdev = &magdev->busdev;
+
+    bool ack = busReadRegisterBuffer(busdev, HMC58X3_REG_DATA, buf, 6);
+
     if (!ack) {
         return false;
     }
     // During calibration, magGain is 1.0, so the read returns normal non-calibrated values.
     // After calibration is done, magGain is set to calculated gain values.
 
-    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magGain[X];
-    magData[Z] = (int16_t)(buf[2] << 8 | buf[3]) * magGain[Z];
-    magData[Y] = (int16_t)(buf[4] << 8 | buf[5]) * magGain[Y];
+    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magdev->magGain[X] / 256;
+    magData[Z] = (int16_t)(buf[2] << 8 | buf[3]) * magdev->magGain[Z] / 256;
+    magData[Y] = (int16_t)(buf[4] << 8 | buf[5]) * magdev->magGain[Y] / 256;
 
     return true;
 }
 
-static bool hmc5883lInit(void)
+static bool hmc5883lInit(magDev_t *magdev)
 {
+    busDevice_t *busdev = &magdev->busdev;
+
     int16_t magADC[3];
     int i;
     int32_t xyz_total[3] = { 0, 0, 0 }; // 32 bit totals so they won't overflow.
     bool bret = true;           // Error indicator
 
+    magdev->magGain[X] = 256;
+    magdev->magGain[Y] = 256;
+    magdev->magGain[Z] = 256;
+
     delay(50);
-#ifdef USE_MAG_SPI_HMC5883
-    hmc5883SpiWriteCommand(HMC58X3_R_CONFA, 0x010 + HMC_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
-#else
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFA, 0x010 + HMC_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
-#endif
+
+    busWriteRegister(busdev, HMC58X3_REG_CONFA, HMC_CONFA_DOR_15HZ | HMC_CONFA_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
+
     // Note that the  very first measurement after a gain change maintains the same gain as the previous setting.
     // The new gain setting is effective from the second measurement and on.
-#ifdef USE_MAG_SPI_HMC5883
-    hmc5883SpiWriteCommand(HMC58X3_R_CONFB, 0x60); // Set the Gain to 2.5Ga (7:5->011)
-#else
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFB, 0x60); // Set the Gain to 2.5Ga (7:5->011)
-#endif
+
+    busWriteRegister(busdev, HMC58X3_REG_CONFB, HMC_CONFB_GAIN_2_5GA); // Set the Gain to 2.5Ga (7:5->011)
+
     delay(100);
-    hmc5883lRead(magADC);
+
+    hmc5883lRead(magdev, magADC);
 
     for (i = 0; i < 10; i++) {  // Collect 10 samples
```

Comment:
There is big overlap in positive/negative measurement code. It would be nice to move it into separate function or maybe unify it using outer loop.

Reply:
Separate function would be better, just for example:
```
  enum {
    polPos, polNeg
  };
  for (int polarity = polPos; polarity <= polNeg; polarity++) {
  switch(polarity) {
  case polPos:
    busWriteRegister(busdev, HMC58X3_REG_CONFA, HMC_CONFA_DOR_15HZ | HMC_CONFA_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
    break;
  case polNeg:
    busWriteRegister(busdev, HMC58X3_REG_CONFA, HMC_CONFA_DOR_15HZ | HMC_CONFA_NEG_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to negative bias.
    break;
  }
  for(int i =0; i< 10; i++) ... 
  ...
  xyz_total[X] += ((polarity == polPos) ? 1 : -1) * magADC[X];